### PR TITLE
[GraphQl] Better pagination support

### PR DESCRIPTION
**Pull request by Reduxx**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #2139
| License       | MIT
| Doc PR        | api-platform/docs#561

This is a WIP Pull Request. It aims to make the pagination support closer to the graphql-relay-js implementation.

Bug Fixes:
pageInfo.endCursor is now the last item of the current page

New Features:
pageInfo.startCursor: new Type for the first item of the current page
queryArguments last and before: the counterpart for first and after

PHPUnit Tests: passing

Behat: 
- all tests: 5 not passing
- GraphQl tests: passing (although there are no tests for startCursor, last and before at the moment)
I haven't figured out why the other tests aren't running properly, I have to check this.

Any help with the tests are appreciated, this is the first time I've ever seen Behat and used prophecies in PHPUnit

I'll open a PR for the docs and finish this up according to the guidelines when I have the time :)

### CODE REVIEW

**Comment by Reduxx** [comment]
The phpstan errors seem to be unrelated to this PR, they are also thrown on the base master branch.
I couldn't figure out why it says phpunit-coverage failed, the tests are all passing and circleci says "0 failures" but it has an exit code of 1.

The new behavior now working as intended. PHPUnit and Behat tests are updated to reflect this.

The docs PR is almost ready, just figuring out the wording and spelling. And then I'll finalize this :)

**Comment by soyuka** [comment]
rebasing should do !

**Comment by Reduxx** [comment]
I've rebased my commits, but unfortunately something doesn't seem right :-/

But at least I've included the docs PR api-platform/docs#561 :)

File: src/Bridge/Doctrine/Orm/Extension/PaginationExtension.php [hunk 1]
```diff
@@ -121,6 +122,19 @@ public function applyToCollection(QueryBuilder $queryBuilder, QueryNameGenerator
...
                 $firstResult = false === $after ? $firstResult : ++$firstResult;
             }
+            if (isset($collectionArgs[$resourceClass]['last'])) {
+                $firstResult = \count($queryBuilder->getQuery()->getArrayResult()) - $collectionArgs[$resourceClass]['last'];
+                $firstResult = 0 > $firstResult ? 0 : $firstResult;
+            }
+            if (isset($collectionArgs[$resourceClass]['before'])) {
+                $before = \base64_decode($collectionArgs[$resourceClass]['before'], true);
+                $firstResult = (int) $before - $itemsPerPage;
+                $firstResult = (false === $before) ? 0 : (int) $firstResult;
```
Comment (alanpoulain) [inline src/Bridge/Doctrine/Orm/Extension/PaginationExtension.php]: The cast is not needed.

File: src/Bridge/Doctrine/Orm/Extension/PaginationExtension.php [hunk 2]
```diff
@@ -121,6 +122,19 @@ public function applyToCollection(QueryBuilder $queryBuilder, QueryNameGenerator
...
+                if (false !== $before && 0 > $firstResult) {
```
Comment (alanpoulain) [inline src/Bridge/Doctrine/Orm/Extension/PaginationExtension.php]: I think you can remove the `false !== $before` part. Since `$firstResult` will be negative only if `$before` is not `false`.

Comment (alanpoulain) [reply src/Bridge/Doctrine/Orm/Extension/PaginationExtension.php]: Is this part tested?

File: src/Bridge/Doctrine/Orm/Extension/PaginationExtension.php [hunk 3]
```diff
@@ -121,6 +122,19 @@ public function applyToCollection(QueryBuilder $queryBuilder, QueryNameGenerator
                 $firstResult = (int) $after;
                 $firstResult = false === $after ? $firstResult : ++$firstResult;
             }
+            if (isset($collectionArgs[$resourceClass]['last'])) {
```
Comment (alanpoulain) [inline src/Bridge/Doctrine/Orm/Extension/PaginationExtension.php]: Are you doing this for the case the `before` is missing? I think we shouldn't manage this case. The `last` should be only managed when a `before` is present.
And worse, you are doing an extra query inside a building of a query.

File: src/GraphQl/Resolver/Factory/CollectionResolverFactory.php [hunk 4]
```diff
@@ -106,21 +106,44 @@ public function __invoke(string $resourceClass = null, string $rootClass = null,
...
             }
+            if (isset($args['before'])) {
+                $before = base64_decode($args['before'], true);
+                if (false === $before) {
+                    throw Error::createLocatedError(sprintf('Cursor %s is invalid', $args['before']), $info->fieldNodes, $info->path);
+                }
+                $offset = (int) $before - $collection->count();
+                $offset = 0 > $offset ? 0 : $offset;
+            }
+            if (isset($args['last']) && !isset($args['before'])) {
```
Comment (alanpoulain) [inline src/GraphQl/Resolver/Factory/CollectionResolverFactory.php]: As I said above, we shouldn't manage this case.

File: src/GraphQl/Resolver/Factory/CollectionResolverFactory.php [hunk 5]
```diff
@@ -106,21 +106,44 @@ public function __invoke(string $resourceClass = null, string $rootClass = null,
...
             if ($collection instanceof PaginatorInterface && ($totalItems = $collection->getTotalItems()) > 0) {
-                $data['pageInfo']['endCursor'] = base64_encode((string) ($totalItems - 1));
                 $data['pageInfo']['hasNextPage'] = $collection->getCurrentPage() !== $collection->getLastPage() && (float) $collection->count() === $collection->getItemsPerPage();
+                $data['pageInfo']['startCursor'] = \base64_encode((string) ($offset));
                 $data['totalCount'] = $totalItems;
+                $collectionLastIndex = $collection->count() - 1;
             }
 
             foreach ($collection as $index => $object) {
+                if ($index === $collectionLastIndex) {
```
Comment (alanpoulain) [inline src/GraphQl/Resolver/Factory/CollectionResolverFactory.php]: Why is this needed? The `$collectionLastIndex` is always the endCursor doesn't it?

File: src/GraphQl/Resolver/Factory/CollectionResolverFactory.php [hunk 6]
```diff
@@ -106,21 +106,44 @@ public function __invoke(string $resourceClass = null, string $rootClass = null,
...
+                    $endCursor = \base64_encode((string) ($index + $offset));
+                }
                 $data['edges'][$index] = [
                     'node' => $this->normalizer->normalize($object, ItemNormalizer::FORMAT, $dataProviderContext),
-                    'cursor' => base64_encode((string) ($index + $offset)),
+                    'cursor' => \base64_encode((string) ($index + $offset)),
                 ];
             }
 
+            if ($collection instanceof PaginatorInterface && ($totalItems = $collection->getTotalItems()) > 0) {
```
Comment (alanpoulain) [inline src/GraphQl/Resolver/Factory/CollectionResolverFactory.php]: It should be moved inside the previous `if`.

File: src/GraphQl/Resolver/Factory/CollectionResolverFactory.php (hunk 5 above)
Comment (Reduxx) [reply src/GraphQl/Resolver/Factory/CollectionResolverFactory.php]: You are completly right, it's not needed at all

File: src/Bridge/Doctrine/Orm/Extension/PaginationExtension.php (hunk 2 above)
Comment (Reduxx) [reply src/Bridge/Doctrine/Orm/Extension/PaginationExtension.php]: This is correct I'll change that.
It is tested in features/graphql/collection.feature starting on line 424

File: src/Bridge/Doctrine/Orm/Extension/PaginationExtension.php (hunk 3 above)
Comment (Reduxx) [reply src/Bridge/Doctrine/Orm/Extension/PaginationExtension.php]: Yeah, that's kinda hacky. It's for the case that before isn't set, so you can get the last items in the correct order. I have to think about another way

Comment (Reduxx) [reply src/Bridge/Doctrine/Orm/Extension/PaginationExtension.php]: I didn't come up with an idea so far :-/ How should I we this?

**Comment by Reduxx** [comment]
No further changes, only updated branch to match base branch

File: src/Bridge/Doctrine/Orm/Extension/PaginationExtension.php (hunk 3 above)
Comment (alanpoulain) [reply src/Bridge/Doctrine/Orm/Extension/PaginationExtension.php]: I think doing backwards pagination in SQL is tricky and you need to know the total rows the first time in order to do it.
The current implementation needs to be changed though because `getArrayResult` can be very big and I think it can eat all the memory. You need to use an SQL count.
The Doctrine Paginator is doing it anyway when you get the total items (see its method `getCountQuery`). You need to do something like this.

**Comment by alanpoulain** [comment]
@Reduxx Could you try to use an SQL count?

**Comment by alanpoulain** [comment]
This PR should be good.
@api-platform/core-team, I have introduced an `ArrayIterator` in order to have pagination information when returning an array, please tell me what you think of it.

**Comment by alanpoulain** [comment]
@soyuka I have cherry-picked the commit (https://github.com/api-platform/core/commit/f91d3d8d75870e54bd74e9ea500b7fa29d92b461) for the limit 0 which uses constants.

File: src/Bridge/Doctrine/Orm/Extension/PaginationExtension.php [hunk 7]
```diff
@@ -267,6 +269,19 @@ private function getPaginationParameter(Request $request, string $parameterName,
...
     }
 
+    private function addCountToContext(QueryBuilder $queryBuilder, array $context): array
+    {
+        if (!($context['graphql'] ?? false)) {
+            return $context;
+        }
+
+        if (isset($context['filters']['last']) && !isset($context['filters']['before'])) {
+            $context['count'] = (new DoctrineOrmPaginator($queryBuilder))->count();
```
Comment (meyerbaptiste) [inline src/Bridge/Doctrine/Orm/Extension/PaginationExtension.php]: Should I assume that partial pagination is not compatible with Graphql?

File: src/DataProvider/ArrayPaginator.php [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    private $firstResult;
+    private $maxResults;
+    private $totalItems;
+
+    public function __construct(array $results, int $firstResult, int $maxResults)
+    {
+        if ($maxResults > 0) {
+            $this->iterator = new \LimitIterator(new \ArrayIterator($results), $firstResult, $maxResults);
+        } else {
+            $this->iterator = new \ArrayIterator([]);
```
Comment (meyerbaptiste) [inline src/DataProvider/ArrayPaginator.php]: `\EmptyIterator`?

File: src/Bridge/Doctrine/Orm/Extension/PaginationExtension.php (hunk 7 above)
Comment (alanpoulain) [reply src/Bridge/Doctrine/Orm/Extension/PaginationExtension.php]: Yes. We could make it compatible later.

File: src/GraphQl/Resolver/Factory/CollectionResolverFactory.php [hunk 9]
```diff
@@ -101,20 +101,39 @@ public function __invoke(string $resourceClass = null, string $rootClass = null,
...
+                    throw Error::createLocatedError(sprintf('Cursor %s is invalid', $args['before']), $info->fieldNodes, $info->path);
+                }
+                $offset = (int) $before - $nbPageItems;
+            }
+            if (isset($args['last']) && !isset($args['before'])) {
+                $offset = $totalItems - $args['last'];
+            }
+            $offset = 0 > $offset ? 0 : $offset;
+
+            $data = ['totalCount' => 0.0, 'edges' => [], 'pageInfo' => ['startCursor' => null, 'endCursor' => null, 'hasNextPage' => false]];
```
Comment (meyerbaptiste) [inline src/GraphQl/Resolver/Factory/CollectionResolverFactory.php]: `0.`

File: src/Bridge/Doctrine/Orm/Extension/PaginationExtension.php (hunk 7 above)
Comment (alanpoulain) [reply src/Bridge/Doctrine/Orm/Extension/PaginationExtension.php]: The code you annotated is only if you are doing backwards pagination though.

**Comment by alanpoulain** [comment]
This PR misses the `hasPreviousPage` field. Please do not merge it yet.

**Comment by alanpoulain** [comment]
`hasPreviousPage` added.

**Comment by alanpoulain** [comment]
Thank @Reduxx :slightly_smiling_face:

**Comment by luca-simonetti** [comment]
any plans to have this released soon?

**Comment by rajasimon** [comment]
Need this now. My after cursor resulting in empty result after five iteration. Why this not in production?
